File: apps/backend/internal/model/budget/budget.go
package budget

import (
	"time"

	"github.com/google/uuid"
)

type Budget struct {
	ID                uuid.UUID `json:"id" db:"id"`
	CategoryID        uuid.UUID `json:"category_id" db:"category_id"`
	MonthlyLimitMinor int64     `json:"monthly_limit_minor" db:"monthly_limit_minor"`
	Currency          string    `json:"currency" db:"currency"`
	CreatedAt         time.Time `json:"created_at" db:"created_at"`
}

type BudgetMemberRole string

const (
	BudgetMemberRoleOwner  BudgetMemberRole = "owner"
	BudgetMemberRoleMember BudgetMemberRole = "member"
)

type BudgetMember struct {
	BudgetID uuid.UUID        `json:"budget_id" db:"budget_id"`
	UserID   string           `json:"user_id" db:"user_id"`
	Role     BudgetMemberRole `json:"role" db:"role"`
}

type BudgetWithMembers struct {
	Budget
	Members []BudgetMember `json:"members"`
}

File: packages/zod/src/budget.ts
import { z } from "zod";

export const ZBudgetMember = z.object({
  budget_id: z.string().uuid(),
  user_id: z.string(),
  role: z.enum(["owner", "member"]),
});

export const ZBudget = z.object({
  id: z.string().uuid(),
  category_id: z.string().uuid(),
  monthly_limit_minor: z.number(),
  currency: z.string(),
  created_at: z.string().datetime(),
});

export const ZBudgetWithMembers = ZBudget.extend({
  members: z.array(ZBudgetMember),
});

export const ZBudgetResponse = ZBudgetWithMembers.extend({
  usage: z.number(),
  usage_percent: z.number(),
  exceeded: z.boolean(),
});

export const ZCreateBudgetRequest = z.object({
  category_id: z.string().uuid(),
  monthly_limit_minor: z.number().positive(),
  currency: z.string().length(3).optional(),
});

export const ZUpdateBudgetRequest = z.object({
  category_id: z.string().uuid(),
  monthly_limit_minor: z.number().positive(),
  currency: z.string().length(3).optional(),
});

export const ZAddMemberRequest = z.object({
  user_id: z.string(),
});
